<a href="https://colab.research.google.com/github/jprice625/test/blob/main/K_Means_Example.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:

import pandas as pd
import numpy as np

In [2]:
HeartRisk=pd.read_csv('/content/HeartRisk.csv')
HeartRisk

PatientID  Weight  Cholesterol  Sex  Age  Diabetic  Prev_Event  \
0       E58TH     163          169    1   49         0           1   
1      V223QW     199          219    0   50         0           0   
2      R180BJ     247          267    1   61         1           0   
3      L127DQ     285          276    1   75         1           1   
4      W236SF     168          185    0   49         0           1   
..        ...     ...          ...  ...  ...       ...         ...   
542    Q174LD     203          227    0   47         1           0   
543    X243FW     174          193    0   43         1           1   
544     B24SX     236          233    1   52         0           1   
545    K119LS     136          118    0   40         0           0   
546    W231BA     222          240    0   58         0           1   

     Exercise_Issues  HeartRate  
0                  1         71  
1                  1         87  
2                  1        108  
3                  1        125  
4                  1         74  
..               ...        ...  
542                1         89  
543                1         76  
544                1        103  
545                0         60  
546                1         97  

[547 rows x 9 columns]

In [3]:
HeartRisk.describe()

Weight  Cholesterol         Sex         Age    Diabetic  \
count  547.000000   547.000000  547.000000  547.000000  547.000000   
mean   201.400366   210.104205    0.513711   55.764168    0.544790   
std     43.189236    44.413497    0.500269   10.347672    0.498446   
min    133.000000   113.000000    0.000000   40.000000    0.000000   
25%    163.000000   177.000000    0.000000   48.000000    0.000000   
50%    196.000000   216.000000    1.000000   53.000000    1.000000   
75%    240.000000   243.500000    1.000000   63.000000    1.000000   
max    285.000000   298.000000    1.000000   81.000000    1.000000   

       Prev_Event  Exercise_Issues   HeartRate  
count  547.000000       547.000000  547.000000  
mean     0.595978         0.575868   88.239488  
std      0.491151         0.494663   18.913028  
min      0.000000         0.000000   58.000000  
25%      0.000000         0.000000   71.000000  
50%      1.000000         1.000000   86.000000  
75%      1.000000         1.000000  105.000000  
max      1.000000         1.000000  125.000000

Two inputs are required for the KMeans method: a DataFrame of numeric attributes and a number of clusters to generate. Because we have an ID column, PatientID, which is both non-numeric and not useful in identifying cluster groupings, we will use column indexing ([1:9]) to exclude the first column of the HeartRisk DataFrame from our analysis. We will add some other, optional arguments to the KMeans() method as well: max_iter limits the number of times R will loop through the data, re-classifying observations into the different clusters seeking for the greatest possible degree of similarity within and between groups. We will set this to 10 iterations. The argument random_state determines how observations are initially selected into groups for classification. We do not want a random state because this makes replication of results difficult, so we will set this to 0. And finally, n_init allows us to set a specific seed number to start the process of classification, making replication of results easier. This can be set to any number, but for consistency we will use 7. Use the “+ Code” button to add a new command prompt to your Notebook. Enter the command below and then execute the code:

In [4]:
from sklearn.cluster import KMeans
HeartKMeans = KMeans(n_clusters=3, n_init=7, max_iter = 10, random_state=0).fit(HeartRisk.iloc[:, np.r_[1:9]])

Now we will use the HeartKMeans model to create a centroid table. This table will hold the size of each cluster, which is the number of observations grouped into each one. It will also list the centers, or means, for each attribute across all clusters. This table is particularly useful when determining what labels you will assign to each cluster. This will be discussed further in the Evaluation section. Two “for” loops in Python are employed here: one to display the sizes of the clusters and one to display the means for the attributes for each cluster. Add the following code to the Console and press Enter. Python numbers clusters starting with 0. The #comments to the right of the code provide additional explanation of each command’s function.

In [5]:
for i in range(len(HeartKMeans.cluster_centers_)):
  print("Cluster", i, "-- Size:", sum(HeartKMeans.labels_ == i))    #Sum up the number of observations in each cluster and print the results
HeartCentroid = [HeartKMeans.feature_names_in_]     #Create an array with the attribute names defining the number of columns in the array
for p in range(len(HeartKMeans.cluster_centers_)):
    HeartCentroid.append(HeartKMeans.cluster_centers_[p])  #Put the means for each attribute into 'k' number of rows in the array, one for each cluster
HeartCentroid = pd.DataFrame(HeartCentroid)    #Change the array into a pandas DataFrame for easier readability
HeartCentroid.columns = HeartCentroid.iloc[0]
HeartCentroid = HeartCentroid[1:]  #Change the DataFrame's first row into column headers composed of the attribute names
HeartCentroid = HeartCentroid.reset_index()     #Re-number the clusters starting from zero after changing the first row to column headers
HeartCentroid = HeartCentroid.drop('index', axis=1)    #Drop the old row numbering
HeartCentroid


Cluster 0 -- Size: 180
Cluster 1 -- Size: 161
Cluster 2 -- Size: 206


0      Weight Cholesterol       Sex        Age  Diabetic Prev_Event  \
0  153.844444  158.033333  0.555556  45.938889  0.422222   0.422222   
1  257.291925  259.322981  0.596273  66.571429  0.782609   0.863354   
2  199.271845  217.135922  0.412621  55.902913  0.466019   0.538835   

0 Exercise_Issues   HeartRate  
0        0.372222   67.416667  
1        0.838509  112.695652  
2        0.548544   87.320388

Finally, we will want to know which insurance customers (PatientIDs) were placed into each cluster. Add a new command prompt to your Notebook using “+ Code” and execute the command below. This will create a new copy of the HeartRisk DataFrame called RiskClusters that will include each observation’s cluster assignment in a new column called Cluster_Number. The rightmost column of this new DataFrame will display the cluster for each observation.

In [ ]:
RiskClusters = HeartRisk      #Creates a copy of the HeartRisk data in a new DataFrame called RiskClusters
RiskClusters['Cluster_Number']=HeartKMeans.labels_        #Adds the cluster number of each observation in RiskCluster
pd.set_option('display.max_rows', None)          #Ensures each observation will be shown in the results
RiskClusters          #Prints out the DataFrame containing each observation and its cluster assignment

PatientID  Weight  Cholesterol  Sex  Age  Diabetic  Prev_Event  \
0       E58TH     163          169    1   49         0           1   
1      V223QW     199          219    0   50         0           0   
2      R180BJ     247          267    1   61         1           0   
3      L127DQ     285          276    1   75         1           1   
4      W236SF     168          185    0   49         0           1   
5      W235YY     187          214    1   55         1           1   
6      X242NL     285          289    1   80         1           1   
7       E59ES     161          181    1   50         0           1   
8       I97KQ     254          277    0   65         1           1   
9      X242QV     184          207    0   54         1           1   
10      B23SW     171          189    1   40         0           0   
11      B24PD     198          224    0   60         1           0   
12      B29JS     205          233    1   56         0           1   
13      I94SX     273          276    1   69         1           1   
14     P162UV     178          203    0   62         1           1   
15     V221GU     153          151    0   44         0           1   
16     V221BU     181          202    0   53         0           0   
17      E54YN     185          207    0   51         0           1   
18      D42EB     189          216    0   52         1           1   
19     N144KN     167          186    0   52         0           1   
20     M131RA     184          211    1   53         0           0   
21     K114TX     185          215    1   59         0           1   
22     S198OR     240          221    1   60         1           0   
23      A14UN     241          243    1   53         0           1   
24      D44UD     146          147    1   47         1           1   
25      I98ZH     205          231    1   54         1           1   
26      G78LR     215          220    1   49         1           1   
27     Z262BB     216          223    1   61         1           1   
28      D46WZ     285          276    0   67         1           1   
29     J107HQ     178          198    0   65         0           0   
30     N146ZZ     241          264    1   51         1           1   
31     P162SV     174          198    1   42         0           0   
32     S199XS     227          252    1   47         1           0   
33     U213XM     171          190    0   43         1           0   
34      H81QA     143          145    1   40         0           0   
35     Z265RY     220          251    0   49         1           1   
36     T206PP     258          240    1   76         0           1   
37     L125XE     143          150    1   51         1           1   
38     Y253TM     192          218    0   59         1           1   
39      H86FP     257          267    0   72         1           1   
40     Q172WV     173          192    0   51         0           0   
41      F68RR     269          263    0   71         1           1   
42      E52BL     187          206    1   63         0           0   
43     Y254BX     142          123    0   52         0           0   
44     X241PU     259          282    1   64         1           1   
45      D43KC     262          280    1   64         1           1   
46      A14XN     203          229    0   51         0           0   
47     M136GP     161          176    1   47         0           1   
48     L123OM     252          251    0   61         1           1   
49     N142EV     192          212    0   65         0           0   
50      C30FJ     264          244    0   63         1           1   
51     Q175IE     255          238    0   64         1           1   
52     M132JV     216          240    1   47         0           0   
53     X243KW     198          223    0   63         0           1   
54     U213NW     163          180    0   40         0           1   
55     N142PB     262          247    0   80         1           1   
56     W234RN 

People sometimes wonder: How many clusters do I need? What is the right value for k? There is no one right answer to these questions; however, there are some techniques that may be helpful in determining a reasonable number of clusters for a given data set. It is important to remember that the ideal number of clusters for each data set will be different depending on the number of attributes, the variability in the values within and among the attributes, and the size of the data set. There are several techniques to help analysts identify an optimal number of clusters for a given data set. These include visualizations such as scatterplots to look for groupings that may appear across correlated attributes, along with algorithmic approaches such as the Elbow Method, Silhouette Method, and Gap Statistics. Ultimately, the right number of clusters for an analysis is the number that yields the greatest insights into the data. We will discuss this more in the next section: Evaluation.

Sonia’s major objective in the hypothetical scenario posed in this chapter was to try to find groupings of heart disease risk among insurance customers. Using the k-Means technique, we have identified three clusters for Sonia, and we can now evaluate their usefulness in addressing her question. We will start by looking at our centroid table. Each data analytics software product will number clusters differently, but no software will apply labels, such as “Low Risk,” to the clusters. This is the job of the data analyst. The meaning of the values in the clusters is up to us to determine.

Because various analytics software products will display k-means clustering results differently, we will examine the clusters based on the values they contain, rather than by assigned cluster number. Examine Figure 6.16. This is a centroid table as displayed in the RStudio software. Look at the cluster on the top row of the table. There are 180 individuals in this cluster. We might label this cluster as “Low Risk Customers.” Why? Primarily because this cluster has the lowest averages for major indicators of heart disease risk. This group has the lowest average weight, cholesterol, and age; lowest incidence of diabetes, previous heart events, and exercise-induced issues; and lowest resting heart rate. Thus, these health insurance customers are probably at the lowest risk of developing heart problems, at least as they relate to the attributes in this data set. Notice that the average in the Sex variable for this cluster is not the lowest in the table. In this attribute, a value of 0 represents Female and 1 represents Male, so a mean of 0.55 indicates that there are more men than women represented in this cluster.

An examination of the Sex variable across all three clusters shows that only one of the three groups has more females than males in it. The cluster containing 206 people has an average Sex of 0.41. This indicates that there are more zeros (0 = female) than ones (1 = male) in this cluster. Filtering of the data set on assigned cluster number and sex reveals that there are 121 females and 85 males in this cluster. We can determine that 121 zeros plus 85 ones divided by 206 people equals an average Sex value of 0.41. Put another way, approximately 41% of people in this cluster are male and 59% are female. This cluster might be labeled as “Moderate Risk Customers,” since the means across all of the attributes are between the values in the other two clusters. To cite two examples: Cholesterol = 217.1, which falls between 158.0 and 259.3 in the other two clusters, and Age = 55.9, which falls between 45.9 and 66.6 in the other two clusters.

Finally, let’s examine the smallest cluster, the one with 161 individuals in it. The risk factors have the highest averages across all attributes in this cluster. Look at the means for the Diabetic attribute. In this cluster, which we might label “High Risk Customers,” the Diabetic value is almost double the values in both of the other clusters. What might we derive from this observation? First, in the low and moderate risk clusters, there are more people who do not have a diabetic condition than those who do. But in this highest risk group, nearly 80% of people have a diabetic condition. We can surmise from this that diabetes is correlated with other heart disease risk factors, and thus, any interventions that may be developed should probably include diagnosis and treatment for this particular issue.

These are just some insights that can be derived from our initial k-Means analysis of this data. You may spot other interesting comparison values.

In [ ]:
HighRiskCustomers = RiskClusters[RiskClusters['Cluster_Number'] == 0]
HighRiskCustomers

PatientID  Weight  Cholesterol  Sex  Age  Diabetic  Prev_Event  \
0       E58TH     163          169    1   49         0           1   
4      W236SF     168          185    0   49         0           1   
7       E59ES     161          181    1   50         0           1   
10      B23SW     171          189    1   40         0           0   
15     V221GU     153          151    0   44         0           1   
19     N144KN     167          186    0   52         0           1   
24      D44UD     146          147    1   47         1           1   
33     U213XM     171          190    0   43         1           0   
34      H81QA     143          145    1   40         0           0   
37     L125XE     143          150    1   51         1           1   
43     Y254BX     142          123    0   52         0           0   
47     M136GP     161          176    1   47         0           1   
54     U213NW     163          180    0   40         0           1   
56     W234RN     139          118    1   44         0           0   
58     P162VV     147          151    1   40         1           1   
69      A14ID     136          146    0   48         0           0   
70     V225XO     154          162    1   40         1           0   
73      B22XL     146          149    1   45         0           1   
77      B22CB     167          184    0   46         1           1   
80     R189RS     136          118    1   49         0           0   
84     O151EK     136          116    1   49         0           0   
88      I93XM     153          164    1   51         0           1   
94      A13VC     159          171    0   47         0           1   
96      B25EO     163          188    0   49         1           0   
100    K118BR     153          163    0   49         1           1   
102    R185OE     135          117    0   52         0           0   
107    P166OZ     171          186    1   45         1           1   
114     C31KU     157          162    0   41         0           1   
117     H84UX     154          155    1   49         0           1   
119     G72WL     152          158    0   47         0           1   
120     F67KG     150          163    0   47         1           1   
121    T203OW     154          167    0   47         1           1   
123     E52YL     173          189    1   46         0           0   
124    L129GI     166          173    0   48         1           0   
125     A16CF     156          168    0   52         0           1   
133    P165EE     135          113    0   50         0           0   
139     H85YO     145          120    1   40         0           0   
150     H82TB     167          190    0   43         1           0   
155    X245CE     145          145    1   47         0           0   
156    Q178ER     138          118    1   45         0           0   
157    X242QB     154          167    1   44         1           0   
159    U217PG     161          177    1   41         1           0   
160    O152VB     152          166    1   46         1           1   
162     G76SP     149          150    1   40         1           1   
164    P163MM     154          167    0   49         1           1   
169    L120LJ     159          167    1   49         1           1   
172     G79ZS     149          159    1   40         0           1   
173    U218FR     161          179    0   43         1           1   
175     A12EB     170          189    1   44         0           0   
176    P162BL     170          186    1   42         1           1   
178     D46NP     156          160    1   49         1           0   
181    S1903R     170          185    1   42         0           0   
182    P166OF     150          153    0   40         1           0   
184    U211EK     146          151    0   41         1           0   
191    U211AA     167          184    0   46         0           0   
192    U211TK     170          190    1   41         1           0   
195    X246BP 